sysbuild: Reduce maximum application size when MCUboot is enabled

nordicjm opened the pull request:
This adds a new optional Kconfig which allows for reducing the available size of an image without needing to adjust the image slot size in dts. The use of this is for sysbuild whereby bootloaders can adjust the maximum size of an image if they e.g. store metadata at the end of an image which is added externally by a tool or outside source.

The initial implementation allows for a sysbuild-configured application which has MCUboot enabled to reduce the maximum image size to account for the TLV and swap data fields which are present - without this patch it is possible to generate a fully signed image which looks to be valid but is impossible to update a device with because MCUboot itself will say that the image is too large.

It is worth noting that this is an estimate, if someone adds additional TLV fields to their application then this will not be included in the reduced file size but it is not really possible to know the exact size the TLV section will be.

Fixes: #58103

nordicjm committed:
west.yml: MCUboot synchronization from upstream

Update Zephyr fork of MCUboot to revision:
  e9131ee8b8ab9306d8f6ad410fd1e3d8a820ce10

Brings following Zephyr relevant fixes:
 - e9131ee8 zephyr: kconfig: Make saving encrypt TLVs depending
   on encryption
 - 4c7942e5 zephyr: Add estimated image footer size to cache in
   sysbuild
 - 26d6423b boot: zephyr: esp32: zephyr port

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
sysbuild: Add pre_image and post_image hooks

Adds hooks that run before and after every image has been
configured

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
sysbuild: images: mcuboot: Make application depend on bootloader

Makes the MCUboot image configuration task run before the
application image configuration, this is so information can be
generated by MCUboot and then used in the application

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
kconfig: Add image size reduction option

Adds a Kconfig option that can be used to reduce the available
slot size of an application, this is useful for things like
bootloaders whereby the bootloader or bootloader configuration
reduces the available flash size that the image can occupy

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
linker: Add flash end ROM offset

Reduces the size of an image in the linker script by using the
ROM end offset Kconfig value

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

57300 commented on share/sysbuild/CMakeLists.txt:
```diff
@@ -35,7 +35,9 @@ get_property(IMAGES GLOBAL PROPERTY sysbuild_images)
 sysbuild_module_call(PRE_CMAKE MODULES ${SYSBUILD_MODULE_NAMES} IMAGES ${IMAGES})
 sysbuild_images_order(IMAGES_CONFIGURATION_ORDER CONFIGURE IMAGES ${IMAGES})
 foreach(image ${IMAGES_CONFIGURATION_ORDER})
+  sysbuild_module_call(PRE_IMAGE_CMAKE MODULES ${SYSBUILD_MODULE_NAMES} IMAGES ${IMAGES} IMAGE ${image})
```
If we go for these hooks, then I believe they should be defined per image, not per module. In other words, instead of a `${SYSBUILD_CURRENT_MODULE_NAME}_pre_image_cmake` function, support having a `${image}_pre_image_cmake`.

nordicjm replied:
I don't agree with that, whilst in this instance it is mcuboot which has the hook, the hook could very well reside in a module, it could be needed for multiple applications not just one, therefore making an image name specific hook is much more cumbersome

57300 replied:
I think it would be more convenient to add a separate hook function for each new image, instead of hunting down and patching an existing, module-specific hook. If multiple images need to share the same hook, then the common code could be moved into some macro, called from multiple functions.

Decoupling this from modules would also open up the possibility of adding hooks for images within Zephyr.

nordicjm replied:
> I think it would be more convenient to add a separate hook function for each new image, instead of hunting down and patching an existing, module-specific hook. If multiple images need to share the same hook, then the common code could be moved into some macro, called from multiple functions.

But you can't then do that, how do you have a single function that runs for **all** images after each one has been configured when you don't know the possible names of the applications or want to accept any application name?

57300 replied:
I see. I can't think of a use case for **all** images having a shared `POST_IMAGE_CMAKE` hook, but if that's important, then using modules is fine.

tejlmand replied:
while `pre` and `post` hooks may be convenient and more visible to users, then I just want to point out the fact that because those hooks surrounds a single function, that is: `ExternalZephyrProject_Cmake(APPLICATION ${image})`
then CMake already provides a mechanism for this.

Any module can redeclare the `ExternalZephyrProject_Cmake(APPLICATION ${image})` function like this, and add their hooks.

For example in Zephyr module `foo` then do this:
```
# foo/sysbuild/CMakeLists.txt

function(ExternalZephyrProject_Cmake)
    # Call any prehook functions.
    prehook_functions()
  
  # Call original implementation.
   _ExternalZephyrProject_Cmake(${ARGN})

  # Call any posthook functions.
  posthook_functions()
endfunction()
```
but as said, I'm fine with an official and documented sysbuild feature for this.

> I think it would be more convenient to add a separate hook function for each new image

but that we have already for the `prehook` part through the IMAGE_CONF_SCRIPT property.
So anywhere before CMake is invoked for the images then  you can do the following:
```
set_property(TARGET <image> APPEND PROPERTY IMAGE_CONF_SCRIPT <myscript.cmake>)
```
Result, before CMake is invoked on `<image>`, then `<myscript.cmake>` is included.

nordicjm replied:
> but that we have already for the prehook part through the IMAGE_CONF_SCRIPT property.

This has a default set for images already which would then prevent the default from running, so it's not really usable like that

57300 commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -481,13 +481,15 @@ endfunction()
-# POST_DOMAINS: Invoke post-domains call for modules after creation of domains yaml.
+# PRE_CMAKE       : Invoke pre-CMake call for modules before CMake configure is invoked for images
+# POST_CMAKE      : Invoke post-CMake call for modules after CMake configure has been invoked for images
+# PRE_IMAGE_CMAKE : Invoke pre-CMake call for modules before CMake configure is invoked for each image
+# POST_IMAGE_CMAKE: Invoke post-CMake call for modules after CMake configure has been invoked for each image
```
Shouldn't these be added before the manifest commit?

nordicjm replied:
Fixed

57300 commented on Kconfig.zephyr:
```diff
@@ -111,6 +111,17 @@ config FLASH_LOAD_OFFSET
 
 	  If unsure, leave at the default value 0.
 
+config FLASH_LOAD_OVERHEAD_SIZE
```
The purpose of this seems similar to that of `ROM_START_OFFSET`. Would it be appropriate to call it something like `ROM_END_OFFSET`?

nordicjm replied:
Fixed

57300 commented on include/zephyr/arch/arm/cortex_a_r/scripts/linker.ld:
Should CMake linker scripts be covered as well?

nordicjm replied:
That they should, apparently I was grep'ing in the wrong folder

nordicjm replied:
Fixed

nordicjm committed:
cmake: linker: arm: Fix missing flash load size use

Fixes a mismatch between the cmake and ld linker scripts whereby
the cmake script was missing using the flash load size Kconfig
option (if it was said), which would result in images far larger
than they should have allowed to be

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented:
>  The use of this is for sysbuild whereby bootloaders can adjust the maximum size of an image if they e.g. store metadata at the end of an image which is added externally by a tool or outside source.

just to clarify, isn't it true that the same thing can happen if you build mcuboot and the application individually (that is, without sysbuild) ?

Problem may just be more visible with sysbuild, as it is now possible to automatically have an app built together with mcuboot.

nordicjm commented:
> > The use of this is for sysbuild whereby bootloaders can adjust the maximum size of an image if they e.g. store metadata at the end of an image which is added externally by a tool or outside source.
> 
> just to clarify, isn't it true that the same thing can happen if you build mcuboot and the application individually (that is, without sysbuild) ?
> 
> Problem may just be more visible with sysbuild, as it is now possible to automatically have an app built together with mcuboot.

Yes, it is always present, it's just that you can fix it when using sysbuild but cannot without it (unless someone manually calculates all the required config sizes and sets the Kconfig manually)

tejlmand commented on share/sysbuild/CMakeLists.txt:
```diff
@@ -35,7 +35,9 @@ get_property(IMAGES GLOBAL PROPERTY sysbuild_images)
 foreach(image ${IMAGES_CONFIGURATION_ORDER})
+  sysbuild_module_call(PRE_IMAGE_CMAKE MODULES ${SYSBUILD_MODULE_NAMES} IMAGES ${IMAGES} IMAGE ${image})
   ExternalZephyrProject_Cmake(APPLICATION ${image})
+  sysbuild_module_call(POST_IMAGE_CMAKE MODULES ${SYSBUILD_MODULE_NAMES} IMAGES ${IMAGES} IMAGE ${image})
```
`IMAGE` doesn't seems to be known / defined argument to `sysbuild_module_call()`, so if you want things to work, then you should either add `IMAGE` to list of known and processed args, or if the `IMAGE` arg is intended for the hook implementation, then update the call to the following:
```suggestion
  sysbuild_module_call(POST_IMAGE_CMAKE MODULES ${SYSBUILD_MODULE_NAMES} IMAGES ${IMAGES} EXTRA_ARGS IMAGE ${image})
```

same comment applies to the pre-hook.

nordicjm replied:
Fixed

tejlmand commented:
Is this also a fix for https://github.com/zephyrproject-rtos/zephyr/issues/27471 ?

nordicjm commented:
> Is this also a fix for #27471 ?

The original bug description in that does not really make any sense, slot0 and slot1 are not part of settings, it doesn't make sense that one of them would overlap settings and cause issues for the other partition.

nordicjm commented:
@d3zd3z @microbuilder @de-nordic any comments on this?

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -471,34 +471,50 @@ function(ExternalZephyrProject_Cmake)
-# PRE_DOMAINS : Invoke pre-domains call for modules before creating domains yaml.
-# POST_DOMAINS: Invoke post-domains call for modules after creation of domains yaml.
+# PRE_CMAKE       : Invoke pre-CMake call for modules before CMake configure is invoked for images
+#                   (IMAGES variable provided)
```
Technically it's the caller of `sysbuild_module_call(PRE_CMAKE ...)` which decides that `IMAGES` variable is provided and not this function.

If you want to have `IMAGES` a required argument for the `sysbuild_module_call(PRE_CMAKE ...)` call, then we should check that `IMAGES` is passed along with `PRE_CMAKE`.

This is mixing documentation of the function itself, with how sysbuild actually uses the function.

Same comment applies to the other hook descriptions.

I'm in favor of making the `IMAGES` or `IMAGE` arguments mandatory to reflect how we actually uses the hooks, which also means the text can reflect this requirement.

nordicjm replied:
have updated the text and made variables mandatory

tejlmand reviewed (dismissed):
looks good.

A minor nit on the function description.
I think we could make the arguments mandatory, something you already indicate in your description update.

Approving as such update can be created as followup.

nordicjm committed:
doc: release: 3.6: Add note on changes with build system

Adds notes on changes introduced with sysbuild

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
@feilongfl @almir-okato @sylvioalves this includes the "boot: zephyr: esp32: zephyr port" commit for MCUboot

nordicjm commented:
@henrikbrixandersen @fabiobaltieri would you be interested in reviewing this feature?

carlescufi commented:
> @henrikbrixandersen self-requested a review

Sorry @henrikbrixandersen I totally missed this before merging.

henrikbrixandersen commented:
> > @henrikbrixandersen self-requested a review
> 
> 
> 
> Sorry @henrikbrixandersen I totally missed this before merging.

No worries. I'll look into this after the fact as it addresses an issue, we were having downstream.
